content/blog/r-mpi-part-two.md

```diff
@@ -0,0 +1,298 @@
+result.
+
+Let's run it. After building and installing the package, I ran it
+locally with (for example) 
+`mpiexec -n 8 Rscript -e 'mpitest:::naive_reduce()'` - 
+and on our cluster using the same script from part 1, but setting 
+`/numnodes` to be the same as the number of processes - one process
+per computer. Running with between 1 and 8 processes:-
+
+![Native MPI performance](/img/mpi_2_eg1.png)
+
+Some things we can notice here:-
+
+* The time taken to do the MPI reduction (blue) increases linearly with
+  processes when run locally. We'd expect this because each process needs
+  800Mb of OS RAM. Doing the reduce as effectively a RAM operation on one
+  computer takes more time as the amount of RAM goes up. The local time (green) 
+  also suffers in the same way, because the OS takes longer to give all the
+  processes the memory they request, as the number of processes increases.
+* On the multi-node runs though, the MPI work is spread across the nodes, so
+  we don't get that degradation we got when we were running everything on one
+  node. The local time (green) also gets better as
+  we add nodes, because we really are doing less work per node. 
+* However, we don't get great _efficiency_ - doubling the number of processes
+  doesn't half the time taken, even when only considering the green line. This
+  is mainly because of [Amdahl's law](https://en.wikipedia.org/wiki/Amdahl%27s_law) - 
+  the time taken to allocate memory is included in my local time and no amount
+  of parallelisation will make that faster. Further, we're not really doing
+  that much work in the loop, so the sequential part of getting the memory
+  is a significant chunk of the total time. That limits how much speed-up
+  we could ever achieve.
+
+# Lightening the load
+
+What if we relax the problem a little. Suppose instead of all processes
+needing all of the results back, only process zero needs the assembled
+bulk. All the other processes could then only allocate memory for the 
+data they create, contribute just that to the MPI call.
+
+The new function looks like this:-
+
+```
+gather <- function() {
+  start_mpi()                                       # Start up MPI
+  rank <- get_mpi_rank()                            # Who am I
+  size <- get_mpi_size()                            # How big is my family?
+  local_time <- -(get_mpi_time())
+
+  n <- (100000000 %/% size) * size                  # Ensure n % size = 0
+
+  each <- n / size
+  start <- 1 + (rank * each)
+  end <- min(start + (each - 1), n)
+
+  send_data <- runif(1 + (end - start))            # My data to send
+  send_ptr <- new_data(send_data)                  # Make C++ data pointer
+
+  recv_data <- if (rank == 0) rep(0, n) else 0     # Only make big array 
+  recv_ptr <- new_data(recv_data)                  # on rank 0
+
+  mpi_time <- -(get_mpi_time())
+  mpi_gather_to_zero(send_ptr, recv_ptr)           # Gather in recv_ptr
```

Comment:
Here, is it the case that no data is actually gathered into the `recv_ptr` for the non-rank-0 processes because they aren't valid pointers to the type of data required? Do you always need to pass a pointer to 0 for the non-gathering processes?

Reply:
It's a funny business really. Yes, for the non-zero nodes, the recv_ptr call is ignored. We need all the processes to reach the MPI_Gather command at the same time, and we basically need to supply a pointer to the recv_ptr just to make that line compile. I think we could alternatively say something like this if we wanted :-
```
mpi_gather_to_zero(send_ptr, (rank == 0) ? recv_ptr : (void *) 0)
```
which is both a bit clearer, and slightly messier at the same time...! We are sort of trusting MPI_Gather to not do anything with the invalid thing we send, if we are a non-gathering node.